api/pkg/s3/bucketlifecycleput.go

```diff
@@ -0,0 +1,135 @@
+	"github.com/micro/go-log"
+
+	"crypto/md5"
+
+	"github.com/opensds/multi-cloud/api/pkg/policy"
+
+	"github.com/opensds/multi-cloud/s3/pkg/model"
+	s3 "github.com/opensds/multi-cloud/s3/proto"
+	"golang.org/x/net/context"
+)
+
+func (s *APIService) BucketLifecyclePut(request *restful.Request, response *restful.Response) {
+	if !policy.Authorize(request, response, "bucket:put") {
+		return
+	}
+	bucketName := request.PathParameter("bucketName")
+	log.Logf("Received request for create bucket: %s", bucketName)
+	ctx := context.Background()
+	bucket, _ := s.s3Client.GetBucket(ctx, &s3.Bucket{Name: bucketName})
+	body := ReadBody(request)
+	log.Logf("Body request is %v\n", body)
+	log.Logf("%x", md5.Sum(body))
+
+	if body != nil {
+		createLifecycleConf := model.LifecycleConfiguration{}
+		log.Logf("Before unmarshal struct is %v\n", createLifecycleConf)
+		err := xml.Unmarshal(body, &createLifecycleConf)
+		log.Logf("After unmarshal struct is %v\n", createLifecycleConf)
+		if err != nil {
+			response.WriteError(http.StatusInternalServerError, err)
+			return
+		} else {
+			s3RulePtrArr := make([]*s3.LifecycleRule, 0)
+			for _, rule := range createLifecycleConf.Rule {
+				s3Rule := s3.LifecycleRule{}
+
+				// assign the fields
+				s3Rule.ID = rule.ID //Assigning the rule ID
+
+				//Assigning the status value to s3 status
+				log.Logf("Status in rule file is %v\n", rule.Status)
+				s3Rule.Status = rule.Status
+
+				//Assigning the filter, using convert function to convert xml struct to s3 struct
+				s3Rule.Filter = convertRuleFilterToS3Filter(rule.Filter)
+
+				// Create the type of transition array
+				s3ActionArr := make([]*s3.Action, 0)
+
+				for _, transition := range rule.Transition {
+
+					//Defining the Transition array and assigning the values tp populate fields
+					s3Transition := s3.Action{Name: "Transition"}
+
+					//Assigning the value of days for transition to happen
+					s3Transition.Days = transition.Days
+
+					//Assigning the backend value to the s3 struct
+					s3Transition.Backend = transition.Backend
+
+					//Assigning the storage class of the object to s3 struct
+					//tier, err := s.class2tier(transition.StorageClass)
+					//if err != nil{
+					//	response.WriteError(http.StatusBadRequest, err)
+					//	return
+					//}
+					//s3Transition.Tier = tier
+
+					//Adding the transition value to the main rule
+					s3ActionArr = append(s3ActionArr, &s3Transition)
+				}
+
+				//Loop for getting the values from xml struct
+				for _, expiration := range rule.Expiration {
+					s3Expiration := s3.Action{Name: "Expiration"}
+					s3Expiration.Days = expiration.Days
+					s3ActionArr = append(s3ActionArr, &s3Expiration)
+				}
+				//Assigning the Expiration action to s3 struct expiration
+				s3Rule.Actions = s3ActionArr
+
+				s3Rule.AbortIncompleteMultipartUpload = convertRuleUploadToS3Upload(rule.AbortIncompleteMultipartUpload)
+				// add to the s3 array
+				s3RulePtrArr = append(s3RulePtrArr, &s3Rule)
```

Comment:
The validation check of rule is necessary, for example, if only expiration action exist, then expiration days should be more than one, if expiration and transition both exist, the expiration days should be more than days of transition for 30 days.

Reply:
Done in the latest commit